File: 8. Heaps/23. Merge k Sorted Lists/Solution.cpp
#include <vector>
#include <tuple>
#include <queue>

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* mergeKLists(std::vector<ListNode*>& lists) {
        typedef std::tuple<int, int, ListNode*> Tuple;
        std::priority_queue<Tuple, std::vector<Tuple>, std::greater<Tuple>> min_heap;
        int i = 0;
        for (ListNode* head : lists)
        {
            if (head)
            {
                min_heap.push(std::make_tuple(head->val, i++, head));
            }
        }

        ListNode* dummy = new ListNode();
        ListNode* curr = dummy;
        while (!min_heap.empty())
        {
            auto [val, i, node] = min_heap.top();
            min_heap.pop();
            curr->next = node;
            curr = curr->next;
            if (node->next)
            {
                min_heap.push(std::make_tuple(node->next->val, i, node->next));
            }
        }

        ListNode* res = dummy->next;
        delete dummy;
        return res;
    }
};

File: 8. Heaps/23. Merge k Sorted Lists/Solution.py
from typing import List, Optional
import heapq

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        min_heap = []
        i = 0
        for head in lists:
            if head:
                heapq.heappush(min_heap, (head.val, i, head))
                i += 1
        
        dummy = ListNode()
        curr = dummy

        while min_heap:
            _, index, node = heapq.heappop(min_heap)
            curr.next = node
            curr = curr.next
            if node.next:
                heapq.heappush(min_heap, (node.next.val, index, node.next))
        
        return dummy.next
    